File: mock-server/src/test/java/com/buyansong/protocol/ImPacketCodecTest.java
package com.buyansong.imserver.protocol;

import org.junit.Test;

import java.nio.charset.StandardCharsets;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;

public class ImPacketCodecTest {
    @Test
    public void encodeThenDecodeKeepsFields() {
        ImPacket packet = new ImPacket(ImCommand.SEND_MESSAGE.value(), "{\"content\":\"hello\"}".getBytes(StandardCharsets.UTF_8));

        ImPacket decoded = ImPacketCodec.decode(ImPacketCodec.encode(packet));

        assertEquals(ImCommand.SEND_MESSAGE.value(), decoded.cmd());
        assertArrayEquals(packet.body(), decoded.body());
    }

    @Test(expected = ProtocolException.class)
    public void rejectsInvalidCrc() {
        byte[] bytes = ImPacketCodec.encode(new ImPacket(ImCommand.HEARTBEAT.value(), "ping".getBytes(StandardCharsets.UTF_8)));
        bytes[bytes.length - 1] = (byte) (bytes[bytes.length - 1] + 1);

        ImPacketCodec.decode(bytes);
    }
}

File: app/src/test/java/com/codex/im/protocol/ImPacketCodecTest.kt
package com.codex.im.protocol

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertThrows
import org.junit.Test

class ImPacketCodecTest {
    @Test
    fun encodeThenDecodeKeepsFields() {
        val packet = ImPacket(cmd = ImCommand.SEND_MESSAGE.value, body = """{"content":"hello"}""".toByteArray())

        val decoded = ImPacketCodec.decode(ImPacketCodec.encode(packet))

        assertEquals(ImCommand.SEND_MESSAGE.value, decoded.cmd)
        assertEquals(ImPacketCodec.VERSION, decoded.version)
        assertArrayEquals(packet.body, decoded.body)
    }

    @Test
    fun rejectsInvalidMagic() {
        val bytes = ImPacketCodec.encode(ImPacket(cmd = ImCommand.HEARTBEAT.value, body = ByteArray(0)))
        bytes[0] = 0x00

        assertThrows(ProtocolException::class.java) {
            ImPacketCodec.decode(bytes)
        }
    }

    @Test
    fun rejectsInvalidLength() {
        val bytes = ImPacketCodec.encode(ImPacket(cmd = ImCommand.HEARTBEAT.value, body = "abc".toByteArray()))
        bytes[6] = 0x10

        assertThrows(ProtocolException::class.java) {
            ImPacketCodec.decode(bytes)
        }
    }

    @Test
    fun rejectsInvalidCrc() {
        val bytes = ImPacketCodec.encode(ImPacket(cmd = ImCommand.HEARTBEAT.value, body = "abc".toByteArray()))
        bytes[bytes.lastIndex] = (bytes.last() + 1).toByte()

        assertThrows(ProtocolException::class.java) {
            ImPacketCodec.decode(bytes)
        }
    }
}
